core/src/main/java/com/linecorp/armeria/common/metric/ServiceNaming.java

```diff
@@ -0,0 +1,88 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+package com.linecorp.armeria.common.metric;
+
+import java.util.Optional;
+
+import com.linecorp.armeria.common.logging.RequestOnlyLog;
+import com.linecorp.armeria.common.util.Unwrappable;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.ServiceConfig;
+
+/**
+ * A naming rule that will be applied to {@link RequestOnlyLog#serviceName()}.
+ */
+@FunctionalInterface
+public interface ServiceNaming {
+
+    /**
+     * Returns the {@link ServiceNaming} which converts into the hard-coded service name.
+     */
+    static ServiceNaming of(String defaultServiceName) {
+        return cfg -> defaultServiceName;
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} which converts into the simple name of given service name.
+     * If {@linkplain ServiceConfig#defaultServiceName() defaultServiceName} is set, strip the package name
+     * from the default service name. If not set, strip the package name from the innermost class name.
+     *
+     * @see Class#getSimpleName()
+     */
+    static ServiceNaming simpleTypeName() {
+        return cfg -> {
+            final String serviceName = Optional.ofNullable(cfg.defaultServiceName()).orElseGet(() -> {
+                Unwrappable unwrappable = cfg.service();
+                while (true) {
+                    final Unwrappable delegate = unwrappable.unwrap();
+                    if (delegate != unwrappable) {
+                        unwrappable = delegate;
+                        continue;
+                    }
+                    return delegate.getClass().getName();
+                }
+            });
+            return serviceName.substring(serviceName.lastIndexOf('.') + 1);
+        };
+    }
+
+    /**
+     * Converts the specified {@linkplain RequestOnlyLog#serviceName() serviceName}
+     * into another service name which is used as a meter tag or distributed trace's span name.
+     *
+     * <p>A naming rule can be set by {@link ServerBuilder#defaultServiceNaming(ServiceNaming)}.
+     * One of pre-defined naming rules is able to be used as follows.
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .serviceNaming(ServiceNaming.simpleTypeName())
+     *   .build()
+     * }</pre>
+     *
+     * <p>If customizing is needed out of given rules, a lambda expression would be applied as follows.
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .serviceNaming(cfg -> {
+     *     return Optional.ofNullable(cfg.defaultServiceName()).orElse("UnknownService");
```

Comment:
My understanding is that `ServiceNaming` is used to determine the value of `ServiceConfig.defaultServiceName()`. How is this even possible?

Reply:
I assumed that `defaultServiceName` is able to be converted when both `defaultServiceName` and `defaultServiceNaming` are set. I will revise it.